scripts/util/sqa.py

```diff
@@ -0,0 +1,245 @@
+    If no in progress TPI found, try to get failed/(in-)error TPIs
+    If no failed TPI found, return None
+    The aborted TPIs are ignored since they don't semantically hold
+    any information about the state of a track
+    """
+    product_versions = _product_versions(charm_name, channel, revision)
+
+    if not product_versions:
+        return None
+
+    passed_test_plan_instances = _joined_test_plan_instances(
+        product_versions, TestPlanInstanceStatus.PASSED
+    )
+    if passed_test_plan_instances:
+        return TestPlanInstanceStatus.PASSED
+
+    in_progress_test_plan_instances = _joined_test_plan_instances(
+        product_versions, TestPlanInstanceStatus.IN_PROGRESS
+    )
+    if in_progress_test_plan_instances:
+        return TestPlanInstanceStatus.IN_PROGRESS
+
+    failed_test_plan_instances = _joined_test_plan_instances(
+        product_versions, TestPlanInstanceStatus.FAILED
+    )
+    if failed_test_plan_instances:
+        return TestPlanInstanceStatus.FAILED
+
+    in_error_test_plan_instances = _joined_test_plan_instances(
+        product_versions, TestPlanInstanceStatus.ERROR
+    )
+    if in_error_test_plan_instances:
+        return TestPlanInstanceStatus.ERROR
+
+    return None
+
+
+def _joined_test_plan_instances(
+    product_versions: list[ProductVersion], status: TestPlanInstanceStatus
+) -> list[UUID]:
+    return [
+        ins
+        for product_version in product_versions
+        for ins in _test_plan_instances(str(product_version.uuid), status)
+    ]
+
+
+def _test_plan_instances(
+    productversion_uuid, status: TestPlanInstanceStatus
+) -> list[UUID]:
+    test_plan_instances_cmd = f"testplaninstance list --format json --productversion-uuid {productversion_uuid} --status '{status.value.lower()}'"
+
+    print(
+        f"Getting test plan instances for product version {productversion_uuid} with status {status}..."
+    )
+    print(test_plan_instances_cmd)
+
+    test_plan_instances_response = _weebl_run(*shlex.split(test_plan_instances_cmd))
+
+    print(json_str := test_plan_instances_response)
+    start_index = json_str.rfind("{")
+
+    if start_index != -1:
+        json_str = json_str[start_index:]
+
+    if not (json_dict := json.loads(json_str.strip())):
+        return []
+
+    uuids = [UUID(item) for item in json_dict[K8S_OPERATOR_TEST_PLAN_NAME]]
+
+    return uuids
+
+
+def _product_versions(charm_name, channel, revision) -> list[ProductVersion]:
+    product_versions_cmd = f"productversion list --name {charm_name} --channel {channel} --revision {revision} --format json"
+
+    print(f"Getting product versions for channel {channel} revision {revision}")
+    print(product_versions_cmd)
+
+    product_versions_response = _weebl_run(*shlex.split(product_versions_cmd))
+   
+    print(product_versions_response)
+    adapter = TypeAdapter(list[ProductVersion])
+    product_versions = adapter.validate_json(product_versions_response.strip())
+    return product_versions
```

Comment:
I feel like i've seen this same parsing more than twice in this file.  Is it time for a function to do that?